src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+            }
+        }
+    }
+
+    fn write(&self, offset: usize, content: &[u8]) -> IoResult<usize> {
+        block_on(self.wait_one(FileTask::Write {
+            offset,
+            bytes: content.to_vec(),
+        }))
+        .map(|x| x.unwrap())
+    }
+
+    fn allocate(&self, offset: usize, size: usize) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Allocate { offset, size })).map(|_| ())
+    }
+
+    async fn wait_one(&self, task: FileTask) -> IoResult<Option<usize>> {
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.disk1.send((task.clone(), cb1)).unwrap();
+        self.disk2.send((task, cb2)).unwrap();
+
+        select! {
+            res1 = f1 => res1.unwrap(),
+            res2 = f2 => res2.unwrap(),
+        }
+    }
+}
+
+impl<F: FileSystem> Drop for HedgedHandle<F> {
+    fn drop(&mut self) {
+        self.disk1.send((FileTask::Stop, Box::new(|_| {}))).unwrap();
+        self.disk2.send((FileTask::Stop, Box::new(|_| {}))).unwrap();
+        self.handle1.take().unwrap().join().unwrap();
+        self.handle2.take().unwrap().join().unwrap();
+    }
+}
+
+impl<F: FileSystem> Handle for HedgedHandle<F> {
+    fn truncate(&self, offset: usize) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Truncate(offset))).map(|_| ())
+    }
+
+    fn file_size(&self) -> IoResult<usize> {
+        block_on(self.wait_one(FileTask::FileSize)).map(|x| x.unwrap())
+    }
+
+    fn sync(&self) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Sync)).map(|_| ())
+    }
+}
+
+pub struct HedgedWriter<F: FileSystem> {
+    disk1: Sender<(FileTask, Callback<usize>)>,
+    disk2: Sender<(FileTask, Callback<usize>)>,
+    writer1: Arc<F::Writer>,
+    writer2: Arc<F::Writer>,
+}
+
+impl<F: FileSystem> HedgedWriter<F> {
+    pub fn new(disk1: Sender<(FileTask, Callback<usize>)>, disk2: Sender<(FileTask, Callback<usize>)>, writer1: Arc<F::Writer>, writer2: Arc<F::Writer>) -> Self {
+      
+        Self {
+            handle,
+
+        }
+    }
+}
+
+impl<F: FileSystem> Write for HedgedWriter<F> {
+    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
+    }
+
+    fn flush(&mut self) -> IoResult<()> {
+    }
+}
+
+impl<F: FileSystem> WriteExt for HedgedWriter<F> {
+    fn truncate(&mut self, offset: usize) -> IoResult<()> {
+    }
+
+    fn allocate(&mut self, offset: usize, size: usize) -> IoResult<()> {
+    }
+}
+
+impl<F: FileSystem> Seek for HedgedWriter<F> {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        match pos {
+            SeekFrom::Start(offset) => self.offset = offset as usize,
+            SeekFrom::Current(i) => self.offset = (self.offset as i64 + i) as usize,
+            SeekFrom::End(i) => self.offset = (self.inner.file_size()? as i64 + i) as usize,
+        }
+        Ok(self.offset as u64)
+    }
+}
+
+pub struct HedgedReader<F: FileSystem> {
+    inner: Arc<HedgedHandle<F>>,
+    offset: usize,
```

Comment:
We don't add any concurrency protection here on the `offset`,  does that mean the offset for the files on two disks are equivalent ?

Reply:
I remember that from the design doc, you mentioned the content of the two disks can be different.  IIUC,  in the case of raft logs being purged on one disk while not yet be purged on the on the other disk,  the `inner: Arc<HedgedHandle<F>>` can points to one file, and points to nothing in another run.

Reply:
No need for concurrency protection, cause it's not `Sync`. Rust compiler would take care of it

Reply:
@Connor1996  could you please reply my 2nd comment?

Reply:
Yes, it's possible to point to one file, and the other points to nothing. Still doesn't matter, if it's purged which means nothing would visit it anymore.

Reply:
The fd may not be meaningful for the disk which is purged, but it's meaningful for the other disk which is not yet purged.